mautic/mautic
# Campaign message queue

[//]: # ( Please answer the following questions: )

| Q  | A
| --- | ---
| Bug fix? | n
| New feature? | y
| Related user documentation PR URL | https://github.com/mautic/documentation/pull/103
| Related developer documentation PR URL | 
| Issues addressed (#s or URLs) | 
| BC breaks? | 
| Deprecations? | 

[//]: # ( Note that all new features should have a related user and/or developer documentation PR in their respective repositories. )

[//]: # ( Required: )
#### Description:
This PR changes the way marketing emails are handled at the moment.  With this, when a campaign email is triggered the email will be sent to a queue to be processed instead of sending it right away.  This is due to the fact of a previous feature introduced in version 2.1.0. Frequency rules for contacts.  

Basically, the way this is handling email sending at the moment, the email is lost if the contact has met its current frequency rules. With this, a message can be rescheduled if a contact was not contacted due to frequency rules, also if a message was failed then, it tries to send again until the max attempts are met or the message was sent successfully.

#### Steps to test this PR:
1. Apply this PR
2. create a new campaign and add a new email to be sent as an action. Make sure you select marketing emails for the queue configuration appears.
3. fill in the priority and retry times. Save your campaign.
4. trigger your campaign and see a new message being inserted in the message_queue table.
5. create this new cron to run every so often and send your messages. like this 
6. This PR also bypasses the transactional email to be sent immediately without checking for contact frequency rules. 

php app/console mautic:campaigns:messagequeue

email messages should be sent or rescheduled, marked as sent, and other data.

[//]: # ( As applicable: )
#### Steps to reproduce the bug:
1. 
2. 

#### List deprecations along with the new alternative:
1. 
2. 

#### List backwards compatibility breaks:
1. 
2.
